#include "mainwindow.h"
#include "./ui_mainwindow.h"
#include <QMessageBox>
#include <QInputDialog>
#include <QFileDialog>
#include <QStandardPaths>
#include <QDesktopServices>
#include <QUrl>
#include <QApplication>
#include <QStyle>
#include <QTimer>
#include <QDateTime>
#include <QJsonDocument>
#include <QJsonObject>
#include <QJsonArray>
#include <QSslConfiguration>
#include <QSslSocket>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , tcpSocket(new QTcpSocket(this))
    , webSocket(new QWebSocket())
    , chatModel(new MessageListModel(this))
    , appSettings(new Settings(this))
    , trayIcon(nullptr)
    , trayMenu(nullptr)
    , connectAction(nullptr)
    , disconnectAction(nullptr)
    , quitAction(nullptr)
    , settingsAction(nullptr)
    , statusLabel(nullptr)
    , connectionStatusLabel(nullptr)
    , progressBar(nullptr)
    , heartbeatTimer(new QTimer(this))
    , reconnectTimer(new QTimer(this))
    , emojiPanel(nullptr)
    , stickerPanel(nullptr)
    , currentUsername("")
    , currentServer("localhost")
    , currentPort(8080)
    , useWebSocket(false)
    , isConnected(false)
    , reconnectAttempts(0)
    , notificationsEnabled(true)
    , soundEnabled(true)
    , theme("dark")
{
    ui->setupUi(this);

    setupUI();
    setupConnections();
    setupTrayIcon();
    setupStatusBar();
    setupEmojiPanel();
    setupStickerPanel();
    loadSettings();

    // Ğ˜Ğ½Ğ¸Ñ†Ğ¸Ğ°Ğ»Ğ¸Ğ·Ğ°Ñ†Ğ¸Ñ Ñ‚Ğ°Ğ¹Ğ¼ĞµÑ€Ğ¾Ğ²
    heartbeatTimer->setInterval(30000); // 30 ÑĞµĞºÑƒĞ½Ğ´
    connect(heartbeatTimer, &QTimer::timeout, this, &MainWindow::sendHeartbeat);

    reconnectTimer->setSingleShot(true);
    reconnectTimer->setInterval(5000); // 5 ÑĞµĞºÑƒĞ½Ğ´
    connect(reconnectTimer, &QTimer::timeout, this, &MainWindow::connectToServer);

    // Ğ—Ğ°Ğ³Ñ€ÑƒĞ·ĞºĞ° ÑÑ‚Ğ¸Ğ»Ñ
    applyTheme(theme);

    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ ÑĞ¾ÑÑ‚Ğ¾ÑĞ½Ğ¸Ñ Ğ¸Ğ½Ñ‚ĞµÑ€Ñ„ĞµĞ¹ÑĞ°
    updateConnectionState();
}

MainWindow::~MainWindow()
{
    if (isConnected) {
        disconnectFromServer();
    }
    delete ui;
}

void MainWindow::setupUI()
{
    // Ğ£ÑÑ‚Ğ°Ğ½Ğ¾Ğ²ĞºĞ° Ğ¼Ğ¾Ğ´ĞµĞ»Ğ¸ Ğ´Ğ»Ñ ÑĞ¿Ğ¸ÑĞºĞ° ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğ¹
    ui->chatView->setModel(chatModel);

    // ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ° Ğ¿Ğ¾Ğ»Ñ Ğ²Ğ²Ğ¾Ğ´Ğ° ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ
    ui->messageInput->setPlaceholderText(tr("Ğ’Ğ²ĞµĞ´Ğ¸Ñ‚Ğµ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğµ..."));

    // ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ° ÑĞ¿Ğ¸ÑĞºĞ° Ğ¿Ğ¾Ğ»ÑŒĞ·Ğ¾Ğ²Ğ°Ñ‚ĞµĞ»ĞµĞ¹ (ĞµÑĞ»Ğ¸ Ğ±ÑƒĞ´ĞµÑ‚ Ñ€ĞµĞ°Ğ»Ğ¸Ğ·Ğ¾Ğ²Ğ°Ğ½)
    ui->contactsList->setVisible(true); // Ğ¢ĞµĞ¿ĞµÑ€ÑŒ Ğ²Ğ¸Ğ´Ğ¸Ğ¼Ğ¾

    // Ğ£ÑÑ‚Ğ°Ğ½Ğ¾Ğ²ĞºĞ° Ğ·Ğ°Ğ³Ğ¾Ğ»Ğ¾Ğ²ĞºĞ° Ğ¾ĞºĞ½Ğ°
    setWindowTitle(tr("ĞœĞµÑÑĞµĞ½Ğ´Ğ¶ĞµÑ€ - ") + appSettings->serverHost() + ":" + QString::number(appSettings->serverPort()));

    // ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ° ĞºĞ½Ğ¾Ğ¿Ğ¾Ğº
    connect(ui->sendButton, &QPushButton::clicked, this, &MainWindow::onSendMessageButtonClicked);
    connect(ui->emojiButton, &QPushButton::clicked, this, &MainWindow::toggleEmojiPanel);
    connect(ui->stickerButton, &QPushButton::clicked, this, &MainWindow::toggleStickerPanel);
    connect(ui->attachButton, &QPushButton::clicked, this, &MainWindow::openAttachmentDialog);

    // ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ° Ğ¿Ğ¾Ğ»Ñ Ğ²Ğ²Ğ¾Ğ´Ğ° ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ
    connect(ui->messageInput, &QLineEdit::returnPressed, this, &MainWindow::onInputReturnPressed);
}

void MainWindow::applyTheme(const QString &themeName)
{
    QString themeFile;
    if (themeName == "dark") {
        themeFile = ":/styles/modern-dark.qss";
    } else if (themeName == "light") {
        themeFile = ":/styles/modern-light.qss";
    } else {
        themeFile = ":/styles/modern-dark.qss"; // Ñ‚ĞµĞ¼Ğ° Ğ¿Ğ¾ ÑƒĞ¼Ğ¾Ğ»Ñ‡Ğ°Ğ½Ğ¸Ñ
    }

    QFile file(themeFile);
    if (file.open(QFile::ReadOnly | QFile::Text)) {
        setStyleSheet(file.readAll());
    }
}

void MainWindow::setupEmojiPanel()
{
    emojiPanel = new QWidget(this);
    emojiPanel->setWindowFlags(Qt::Popup);
    emojiPanel->setFixedSize(350, 400);

    QVBoxLayout *layout = new QVBoxLayout(emojiPanel);

    // Ğ¡Ğ¾Ğ·Ğ´Ğ°ĞµĞ¼ ÑĞµÑ‚ĞºÑƒ ÑĞ¼Ğ¾Ğ´Ğ·Ğ¸
    QGridLayout *emojiGrid = new QGridLayout();
    emojiGrid->setSpacing(5);

    // Ğ‘Ğ°Ğ·Ğ¾Ğ²Ñ‹Ğµ ÑĞ¼Ğ¾Ğ´Ğ·Ğ¸
    QStringList emojis = {"ğŸ˜€", "ğŸ˜‚", "ğŸ˜", "ğŸ¥°", "ğŸ˜", "ğŸ¤©", "ğŸ¥³", "ğŸ˜­", "ğŸ˜¡", "ğŸ¤¯",
                          "ğŸ‘", "ğŸ‘", "ğŸ‘", "ğŸ™Œ", "ğŸ‘Œ", "âœŒï¸", "ğŸ¤", "ğŸ¤Ÿ", "ğŸ¤˜", "ğŸ¤™",
                          "ğŸ‘‹", "ğŸ’ª", "ğŸ¦¾", "â¤ï¸", "ğŸ’–", "ğŸ’˜", "ğŸ’", "ğŸ’“", "ğŸ’", "ğŸ’•",
                          "ğŸ”¥", "â­", "âœ¨", "ğŸ‰", "ğŸ", "ğŸˆ", "ğŸ‚", "ğŸ°", "ğŸ•", "ğŸ”"};

    int rows = 8;
    int cols = 5;

    for (int i = 0; i < emojis.size() && i < rows * cols; ++i) {
        int row = i / cols;
        int col = i % cols;

        QPushButton *emojiBtn = new QPushButton(emojis[i]);
        emojiBtn->setFixedSize(50, 50);
        emojiBtn->setStyleSheet(R"(
            QPushButton {
                font-size: 24px;
                border: none;
                border-radius: 8px;
                background-color: transparent;
            }
            QPushButton:hover {
                background-color: #f0f0f0;
                border: 1px solid #cccccc;
            }
        )");

        connect(emojiBtn, &QPushButton::clicked, this, [this, emoji = emojis[i]]() {
            insertEmoji(emoji);
        });

        emojiGrid->addWidget(emojiBtn, row, col);
    }

    layout->addLayout(emojiGrid);

    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ÑĞµĞ¼ ĞºĞ½Ğ¾Ğ¿ĞºÑƒ Ğ·Ğ°ĞºÑ€Ñ‹Ñ‚Ğ¸Ñ
    QPushButton *closeBtn = new QPushButton(tr("Ğ—Ğ°ĞºÑ€Ñ‹Ñ‚ÑŒ"));
    connect(closeBtn, &QPushButton::clicked, emojiPanel, &QWidget::hide);
    layout->addWidget(closeBtn);

    emojiPanel->hide();
}

void MainWindow::setupStickerPanel()
{
    stickerPanel = new QWidget(this);
    stickerPanel->setWindowFlags(Qt::Popup);
    stickerPanel->setFixedSize(400, 300);

    QVBoxLayout *layout = new QVBoxLayout(stickerPanel);

    // Ğ—Ğ°Ğ³Ğ¾Ğ»Ğ¾Ğ²Ğ¾Ğº
    QLabel *titleLabel = new QLabel(tr("Ğ¡Ñ‚Ğ¸ĞºĞµÑ€Ğ¿Ğ°Ğº"));
    titleLabel->setAlignment(Qt::AlignCenter);
    titleLabel->setStyleSheet("font-size: 16px; font-weight: bold; margin: 10px;");
    layout->addWidget(titleLabel);

    // Ğ¡Ğ¾Ğ·Ğ´Ğ°ĞµĞ¼ ÑĞµÑ‚ĞºÑƒ ÑÑ‚Ğ¸ĞºĞµÑ€Ğ¾Ğ²
    QGridLayout *stickerGrid = new QGridLayout();
    stickerGrid->setSpacing(10);

    // Ğ˜ÑĞ¿Ğ¾Ğ»ÑŒĞ·ÑƒĞµĞ¼ ÑÑ‚Ğ°Ğ½Ğ´Ğ°Ñ€Ñ‚Ğ½Ñ‹Ğµ ÑĞ¼Ğ¾Ğ´Ğ·Ğ¸ ĞºĞ°Ğº "ÑÑ‚Ğ¸ĞºĞµÑ€Ñ‹" Ğ´Ğ»Ñ Ğ´ĞµĞ¼Ğ¾Ğ½ÑÑ‚Ñ€Ğ°Ñ†Ğ¸Ğ¸
    QStringList stickers = {"ğŸ‘»", "â¤ï¸", "â­", "ğŸ”¥", "ğŸ‘", "ğŸ‘", "ğŸ‰", "ğŸ’¯"};

    int rows = 2;
    int cols = 4;

    for (int i = 0; i < stickers.size() && i < rows * cols; ++i) {
        int row = i / cols;
        int col = i % cols;

        QPushButton *stickerBtn = new QPushButton(stickers[i]);
        stickerBtn->setFixedSize(70, 70);
        stickerBtn->setStyleSheet(R"(
            QPushButton {
                font-size: 32px;
                border: 2px solid transparent;
                border-radius: 12px;
                background-color: transparent;
            }
            QPushButton:hover {
                border: 2px solid #0078d4;
                background-color: #f0f8ff;
            }
        )");

        connect(stickerBtn, &QPushButton::clicked, this, [this, sticker = stickers[i]]() {
            insertSticker(sticker);
        });

        stickerGrid->addWidget(stickerBtn, row, col);
    }

    layout->addLayout(stickerGrid);

    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ÑĞµĞ¼ ĞºĞ½Ğ¾Ğ¿ĞºÑƒ Ğ·Ğ°ĞºÑ€Ñ‹Ñ‚Ğ¸Ñ
    QPushButton *closeBtn = new QPushButton(tr("Ğ—Ğ°ĞºÑ€Ñ‹Ñ‚ÑŒ"));
    connect(closeBtn, &QPushButton::clicked, stickerPanel, &QWidget::hide);
    layout->addWidget(closeBtn);

    stickerPanel->hide();
}

void MainWindow::insertEmoji(const QString &emoji)
{
    ui->messageInput->insert(emoji);
    ui->messageInput->setFocus();
}

void MainWindow::insertSticker(const QString &sticker)
{
    // ĞÑ‚Ğ¿Ñ€Ğ°Ğ²Ğ»ÑĞµĞ¼ ÑÑ‚Ğ¸ĞºĞµÑ€ ĞºĞ°Ğº ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğµ
    QString message = QString(":%1:").arg(sticker);

    QJsonObject jsonMsg;
    jsonMsg["type"] = "message";
    jsonMsg["user"] = currentUsername;
    jsonMsg["text"] = message;
    jsonMsg["timestamp"] = QDateTime::currentDateTimeUtc().toString(Qt::ISODateWithMs);
    jsonMsg["message_type"] = "sticker";

    QString jsonString = QJsonDocument(jsonMsg).toJson(QJsonDocument::Compact);

    if (useWebSocket) {
        webSocket->sendTextMessage(jsonString);
    } else {
        tcpSocket->write(jsonString.toUtf8() + "\n");
    }

    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ĞµĞ½Ğ¸Ğµ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ Ğ² Ñ‡Ğ°Ñ‚ ĞºĞ°Ğº ÑĞ¾Ğ±ÑÑ‚Ğ²ĞµĞ½Ğ½Ğ¾Ğµ
    addMessageToChat(currentUsername, message, QDateTime::currentDateTime(), true);

    // Ğ—Ğ°ĞºÑ€Ñ‹Ğ²Ğ°ĞµĞ¼ Ğ¿Ğ°Ğ½ĞµĞ»ÑŒ ÑÑ‚Ğ¸ĞºĞµÑ€Ğ¾Ğ²
    if (stickerPanel->isVisible()) {
        stickerPanel->hide();
    }
}

void MainWindow::toggleEmojiPanel()
{
    if (emojiPanel->isVisible()) {
        emojiPanel->hide();
    } else {
        // ĞŸĞ¾Ğ·Ğ¸Ñ†Ğ¸Ğ¾Ğ½Ğ¸Ñ€ÑƒĞµĞ¼ Ğ¿Ğ°Ğ½ĞµĞ»ÑŒ Ñ€ÑĞ´Ğ¾Ğ¼ Ñ Ğ¿Ğ¾Ğ»ĞµĞ¼ Ğ²Ğ²Ğ¾Ğ´Ğ°
        QPoint pos = ui->messageInput->mapToGlobal(QPoint(0, -emojiPanel->height()));
        emojiPanel->move(pos);
        emojiPanel->show();
    }
}

void MainWindow::toggleStickerPanel()
{
    if (stickerPanel->isVisible()) {
        stickerPanel->hide();
    } else {
        // ĞŸĞ¾Ğ·Ğ¸Ñ†Ğ¸Ğ¾Ğ½Ğ¸Ñ€ÑƒĞµĞ¼ Ğ¿Ğ°Ğ½ĞµĞ»ÑŒ Ñ€ÑĞ´Ğ¾Ğ¼ Ñ Ğ¿Ğ¾Ğ»ĞµĞ¼ Ğ²Ğ²Ğ¾Ğ´Ğ°
        QPoint pos = ui->messageInput->mapToGlobal(QPoint(0, -stickerPanel->height()));
        stickerPanel->move(pos);
        stickerPanel->show();
    }
}

void MainWindow::openAttachmentDialog()
{
    QString fileName = QFileDialog::getOpenFileName(this, tr("Ğ’Ñ‹Ğ±ĞµÑ€Ğ¸Ñ‚Ğµ Ñ„Ğ°Ğ¹Ğ»"), "",
        tr("Ğ˜Ğ·Ğ¾Ğ±Ñ€Ğ°Ğ¶ĞµĞ½Ğ¸Ñ (*.png *.xpm *.jpg *.jpeg);;Ğ”Ğ¾ĞºÑƒĞ¼ĞµĞ½Ñ‚Ñ‹ (*.pdf *.doc *.docx *.txt);;Ğ’ÑĞµ Ñ„Ğ°Ğ¹Ğ»Ñ‹ (*)"));

    if (!fileName.isEmpty()) {
        // Ğ—Ğ´ĞµÑÑŒ Ğ´Ğ¾Ğ»Ğ¶Ğ½Ğ° Ğ±Ñ‹Ñ‚ÑŒ Ñ€ĞµĞ°Ğ»Ğ¸Ğ·Ğ°Ñ†Ğ¸Ñ Ğ¾Ñ‚Ğ¿Ñ€Ğ°Ğ²ĞºĞ¸ Ñ„Ğ°Ğ¹Ğ»Ğ°
        // ĞŸĞ¾ĞºĞ° Ğ¿Ñ€Ğ¾ÑÑ‚Ğ¾ Ğ¿Ğ¾ĞºĞ°Ğ·Ñ‹Ğ²Ğ°ĞµĞ¼ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğµ
        QMessageBox::information(this, tr("Ğ¤Ğ°Ğ¹Ğ» Ğ²Ñ‹Ğ±Ñ€Ğ°Ğ½"),
            tr("Ğ¤Ğ°Ğ¹Ğ»: %1 Ğ±ÑƒĞ´ĞµÑ‚ Ğ¾Ñ‚Ğ¿Ñ€Ğ°Ğ²Ğ»ĞµĞ½").arg(QFileInfo(fileName).fileName()));
    }
}

void MainWindow::setupConnections()
{
    // ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ Ğ´Ğ»Ñ TCP ÑĞ¾ĞºĞµÑ‚Ğ°
    connect(tcpSocket, &QTcpSocket::connected, this, &MainWindow::onSocketConnected);
    connect(tcpSocket, &QTcpSocket::disconnected, this, &MainWindow::onSocketDisconnected);
    connect(tcpSocket, QOverload<QAbstractSocket::SocketError>::of(&QAbstractSocket::error),
            this, &MainWindow::onSocketError);
    connect(tcpSocket, &QTcpSocket::readyRead, this, &MainWindow::onSocketReadyRead);
    
    // ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ Ğ´Ğ»Ñ WebSocket
    connect(webSocket, &QWebSocket::connected, this, &MainWindow::onWebSocketConnected);
    connect(webSocket, &QWebSocket::disconnected, this, &MainWindow::onWebSocketDisconnected);
    connect(webSocket, QOverload<QAbstractSocket::SocketError>::of(&QWebSocket::error),
            this, &MainWindow::onWebSocketError);
    connect(webSocket, &QWebSocket::textMessageReceived,
            this, &MainWindow::onWebSocketTextMessageReceived);
    
    // ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ Ğ´Ğ»Ñ ĞºĞ½Ğ¾Ğ¿Ğ¾Ğº
    connect(ui->connectButton, &QPushButton::clicked, this, &MainWindow::toggleConnection);
    connect(ui->sendButton, &QPushButton::clicked, this, &MainWindow::onSendMessageButtonClicked);
    
    // ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ğµ ÑĞ¸Ğ³Ğ½Ğ°Ğ»Ğ° Ğ½Ğ°Ğ¶Ğ°Ñ‚Ğ¸Ñ Enter Ğ² Ğ¿Ğ¾Ğ»Ğµ Ğ²Ğ²Ğ¾Ğ´Ğ°
    connect(ui->messageInput, &QLineEdit::returnPressed, this, &MainWindow::onInputReturnPressed);
    
    // ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ Ğ´Ğ»Ñ Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº
    connect(appSettings, &Settings::usernameChanged, this, &MainWindow::onUsernameChanged);
    connect(appSettings, &Settings::serverHostChanged, this, &MainWindow::onServerAddressChanged);
    connect(appSettings, &Settings::serverPortChanged, this, &MainWindow::onPortChanged);
    connect(appSettings, &Settings::useSslChanged, this, &MainWindow::onUseSslChanged);
}

void MainWindow::setupTrayIcon()
{
    if (!QSystemTrayIcon::isSystemTrayAvailable()) {
        qWarning() << "Ğ¡Ğ¸ÑÑ‚ĞµĞ¼Ğ½Ñ‹Ğ¹ Ñ‚Ñ€ĞµĞ¹ Ğ½ĞµĞ´Ğ¾ÑÑ‚ÑƒĞ¿ĞµĞ½!";
        return;
    }
    
    trayIcon = new QSystemTrayIcon(this);
    trayIcon->setIcon(QIcon(":/icons/messenger.png")); // ĞŸÑ€ĞµĞ´Ğ¿Ğ¾Ğ»Ğ°Ğ³Ğ°ĞµÑ‚ÑÑ Ğ½Ğ°Ğ»Ğ¸Ñ‡Ğ¸Ğµ Ğ¸ĞºĞ¾Ğ½ĞºĞ¸
    trayIcon->setToolTip(tr("ĞœĞµÑÑĞµĞ½Ğ´Ğ¶ĞµÑ€"));
    
    trayMenu = new QMenu(this);
    
    connectAction = new QAction(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡Ğ¸Ñ‚ÑŒÑÑ"), this);
    disconnectAction = new QAction(tr("ĞÑ‚ĞºĞ»ÑÑ‡Ğ¸Ñ‚ÑŒÑÑ"), this);
    settingsAction = new QAction(tr("ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ¸"), this);
    quitAction = new QAction(tr("Ğ’Ñ‹Ñ…Ğ¾Ğ´"), this);
    
    connect(connectAction, &QAction::triggered, this, &MainWindow::connectToServer);
    connect(disconnectAction, &QAction::triggered, this, &MainWindow::disconnectFromServer);
    connect(settingsAction, &QAction::triggered, this, &MainWindow::showSettingsDialog);
    connect(quitAction, &QAction::triggered, qApp, &QApplication::quit);
    
    trayMenu->addAction(connectAction);
    trayMenu->addAction(disconnectAction);
    trayMenu->addSeparator();
    trayMenu->addAction(settingsAction);
    trayMenu->addSeparator();
    trayMenu->addAction(quitAction);
    
    trayIcon->setContextMenu(trayMenu);
    
    connect(trayIcon, &QSystemTrayIcon::activated, this, &MainWindow::trayActivated);
    
    trayIcon->show();
}

void MainWindow::setupStatusBar()
{
    statusLabel = new QLabel(tr("Ğ“Ğ¾Ñ‚Ğ¾Ğ²"));
    connectionStatusLabel = new QLabel(tr("ĞÑ‚ĞºĞ»ÑÑ‡ĞµĞ½"));
    progressBar = new QProgressBar();
    progressBar->setVisible(false);
    
    statusBar()->addWidget(statusLabel);
    statusBar()->addPermanentWidget(connectionStatusLabel);
    statusBar()->addPermanentWidget(progressBar);
}

void MainWindow::loadSettings()
{
    currentUsername = appSettings->username();
    currentServer = appSettings->serverHost();
    currentPort = appSettings->serverPort();
    useWebSocket = appSettings->useWebSocket();
    notificationsEnabled = appSettings->notificationsEnabled();
    soundEnabled = appSettings->soundEnabled();
    theme = appSettings->theme();
    
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ UI
    ui->usernameEdit->setText(currentUsername);
    ui->serverEdit->setText(currentServer);
    ui->portSpinBox->setValue(currentPort);
    ui->protocolCombo->setCurrentIndex(useWebSocket ? 1 : 0);
    ui->sslCheckBox->setChecked(appSettings->useSsl());
    ui->notificationsCheckBox->setChecked(notificationsEnabled);
    ui->soundCheckBox->setChecked(soundEnabled);
}

void MainWindow::saveSettings()
{
    appSettings->setUsername(currentUsername);
    appSettings->setServerHost(currentServer);
    appSettings->setServerPort(currentPort);
    appSettings->setUseWebSocket(useWebSocket);
    appSettings->setUseSsl(ui->sslCheckBox->isChecked());
    appSettings->setNotificationsEnabled(ui->notificationsCheckBox->isChecked());
    appSettings->setSoundEnabled(ui->soundCheckBox->isChecked());
    appSettings->setTheme(theme);
}

void MainWindow::connectToServer()
{
    if (currentUsername.isEmpty()) {
        currentUsername = QInputDialog::getText(this, tr("Ğ˜Ğ¼Ñ Ğ¿Ğ¾Ğ»ÑŒĞ·Ğ¾Ğ²Ğ°Ñ‚ĞµĞ»Ñ"), tr("Ğ’Ğ²ĞµĞ´Ğ¸Ñ‚Ğµ Ğ²Ğ°ÑˆĞµ Ğ¸Ğ¼Ñ:"));
        if (currentUsername.isEmpty()) {
            showError(tr("Ğ˜Ğ¼Ñ Ğ¿Ğ¾Ğ»ÑŒĞ·Ğ¾Ğ²Ğ°Ñ‚ĞµĞ»Ñ Ğ¾Ğ±ÑĞ·Ğ°Ñ‚ĞµĞ»ÑŒĞ½Ğ¾!"));
            return;
        }
        appSettings->setUsername(currentUsername);
    }
    
    isConnected = false;
    updateConnectionState();
    
    if (useWebSocket) {
        QString scheme = appSettings->useSsl() ? "wss://" : "ws://";
        QString url = QString("%1%2:%3/ws").arg(scheme, currentServer, QString::number(currentPort));
        webSocket->open(QUrl(url));
    } else {
        tcpSocket->connectToHost(currentServer, currentPort);
        
        if (appSettings->useSsl()) {
            // ĞĞ°ÑÑ‚Ñ€Ğ¾Ğ¹ĞºĞ° SSL Ğ´Ğ»Ñ TCP ÑĞ¾ĞºĞµÑ‚Ğ°
            QSslSocket *sslSocket = qobject_cast<QSslSocket*>(tcpSocket);
            if (sslSocket) {
                sslSocket->connectToHostEncrypted(currentServer, currentPort);
            }
        }
    }
    
    statusLabel->setText(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ğµ..."));
}

void MainWindow::disconnectFromServer()
{
    heartbeatTimer->stop();
    
    if (useWebSocket) {
        webSocket->close();
    } else {
        tcpSocket->disconnectFromHost();
    }
    
    isConnected = false;
    updateConnectionState();
    statusLabel->setText(tr("ĞÑ‚ĞºĞ»ÑÑ‡ĞµĞ½"));
}

void MainWindow::toggleConnection()
{
    if (isConnected) {
        disconnectFromServer();
    } else {
        connectToServer();
    }
}

void MainWindow::sendMessage()
{
    QString message = ui->messageInput->text().trimmed();
    if (message.isEmpty()) {
        return;
    }
    
    QJsonObject jsonMsg;
    jsonMsg["type"] = "message";
    jsonMsg["user"] = currentUsername;
    jsonMsg["text"] = message;
    jsonMsg["timestamp"] = QDateTime::currentDateTimeUtc().toString(Qt::ISODateWithMs);
    
    QString jsonString = QJsonDocument(jsonMsg).toJson(QJsonDocument::Compact);
    
    if (useWebSocket) {
        webSocket->sendTextMessage(jsonString);
    } else {
        tcpSocket->write(jsonString.toUtf8() + "\n");
    }
    
    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ĞµĞ½Ğ¸Ğµ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ Ğ² Ñ‡Ğ°Ñ‚ ĞºĞ°Ğº ÑĞ¾Ğ±ÑÑ‚Ğ²ĞµĞ½Ğ½Ğ¾Ğµ
    addMessageToChat(currentUsername, message, QDateTime::currentDateTime(), true);
    
    // ĞÑ‡Ğ¸ÑÑ‚ĞºĞ° Ğ¿Ğ¾Ğ»Ñ Ğ²Ğ²Ğ¾Ğ´Ğ°
    ui->messageInput->clear();
}

void MainWindow::onSendMessageButtonClicked()
{
    sendMessage();
}

void MainWindow::onInputReturnPressed()
{
    sendMessage();
}

void MainWindow::onSocketConnected()
{
    statusLabel->setText(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½ Ğº ÑĞµÑ€Ğ²ĞµÑ€Ñƒ"));
    isConnected = true;
    updateConnectionState();
    
    // ĞÑ‚Ğ¿Ñ€Ğ°Ğ²ĞºĞ° Ğ·Ğ°Ğ¿Ñ€Ğ¾ÑĞ° Ğ°ÑƒÑ‚ĞµĞ½Ñ‚Ğ¸Ñ„Ğ¸ĞºĞ°Ñ†Ğ¸Ğ¸
    sendAuthRequest();
    
    // Ğ—Ğ°Ğ¿ÑƒÑĞº Ñ‚Ğ°Ğ¹Ğ¼ĞµÑ€Ğ° Ğ´Ğ»Ñ Ğ¿Ñ€Ğ¾Ğ²ĞµÑ€ĞºĞ¸ ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ñ
    heartbeatTimer->start();
    
    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ĞµĞ½Ğ¸Ğµ ÑĞ¸ÑÑ‚ĞµĞ¼Ğ½Ğ¾Ğ³Ğ¾ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ
    addMessageToChat(tr("Ğ¡Ğ¸ÑÑ‚ĞµĞ¼Ğ°"), tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¾ Ğº ÑĞµÑ€Ğ²ĞµÑ€Ñƒ"), QDateTime::currentDateTime(), false);
}

void MainWindow::onSocketDisconnected()
{
    statusLabel->setText(tr("ĞÑ‚ĞºĞ»ÑÑ‡ĞµĞ½ Ğ¾Ñ‚ ÑĞµÑ€Ğ²ĞµÑ€Ğ°"));
    isConnected = false;
    updateConnectionState();
    heartbeatTimer->stop();
    
    addMessageToChat(tr("Ğ¡Ğ¸ÑÑ‚ĞµĞ¼Ğ°"), tr("Ğ¡Ğ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ğµ Ğ¿Ğ¾Ñ‚ĞµÑ€ÑĞ½Ğ¾"), QDateTime::currentDateTime(), false);
    
    // ĞŸĞ¾Ğ¿Ñ‹Ñ‚ĞºĞ° Ğ¿ĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ
    if (appSettings->autoReconnect() && reconnectAttempts < 5) {
        reconnectAttempts++;
        statusLabel->setText(tr("ĞŸĞ¾Ğ¿Ñ‹Ñ‚ĞºĞ° Ğ¿ĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ... (%1)").arg(reconnectAttempts));
        reconnectTimer->start();
    }
}

void MainWindow::onSocketError(QAbstractSocket::SocketError error)
{
    QString errorMsg;
    switch (error) {
    case QAbstractSocket::ConnectionRefusedError:
        errorMsg = tr("Ğ¡ĞµÑ€Ğ²ĞµÑ€ Ğ¾Ñ‚ĞºĞ»Ğ¾Ğ½Ğ¸Ğ» ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ğµ");
        break;
    case QAbstractSocket::RemoteHostClosedError:
        errorMsg = tr("Ğ¡ĞµÑ€Ğ²ĞµÑ€ Ğ·Ğ°ĞºÑ€Ñ‹Ğ» ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ğµ");
        break;
    case QAbstractSocket::HostNotFoundError:
        errorMsg = tr("Ğ¡ĞµÑ€Ğ²ĞµÑ€ Ğ½Ğµ Ğ½Ğ°Ğ¹Ğ´ĞµĞ½");
        break;
    case QAbstractSocket::SocketTimeoutError:
        errorMsg = tr("Ğ¢Ğ°Ğ¹Ğ¼Ğ°ÑƒÑ‚ ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ñ");
        break;
    default:
        errorMsg = tcpSocket->errorString();
        break;
    }
    
    showError(errorMsg);
    statusLabel->setText(tr("ĞÑˆĞ¸Ğ±ĞºĞ°: %1").arg(errorMsg));
}

void MainWindow::onSocketReadyRead()
{
    while (tcpSocket->canReadLine()) {
        QByteArray line = tcpSocket->readLine();
        QString messageStr = QString::fromUtf8(line).trimmed();
        
        QJsonDocument doc = QJsonDocument::fromJson(messageStr.toUtf8());
        if (!doc.isNull()) {
            QJsonObject jsonObj = doc.object();
            handleIncomingMessage(jsonObj);
        }
    }
}

void MainWindow::onWebSocketConnected()
{
    statusLabel->setText(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½ Ğº ÑĞµÑ€Ğ²ĞµÑ€Ñƒ (WebSocket)"));
    isConnected = true;
    updateConnectionState();
    
    // ĞÑ‚Ğ¿Ñ€Ğ°Ğ²ĞºĞ° Ğ·Ğ°Ğ¿Ñ€Ğ¾ÑĞ° Ğ°ÑƒÑ‚ĞµĞ½Ñ‚Ğ¸Ñ„Ğ¸ĞºĞ°Ñ†Ğ¸Ğ¸
    sendAuthRequest();
    
    // Ğ—Ğ°Ğ¿ÑƒÑĞº Ñ‚Ğ°Ğ¹Ğ¼ĞµÑ€Ğ° Ğ´Ğ»Ñ Ğ¿Ñ€Ğ¾Ğ²ĞµÑ€ĞºĞ¸ ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ñ
    heartbeatTimer->start();
    
    // Ğ”Ğ¾Ğ±Ğ°Ğ²Ğ»ĞµĞ½Ğ¸Ğµ ÑĞ¸ÑÑ‚ĞµĞ¼Ğ½Ğ¾Ğ³Ğ¾ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ
    addMessageToChat(tr("Ğ¡Ğ¸ÑÑ‚ĞµĞ¼Ğ°"), tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¾ Ğº ÑĞµÑ€Ğ²ĞµÑ€Ñƒ (WebSocket)"), QDateTime::currentDateTime(), false);
}

void MainWindow::onWebSocketDisconnected()
{
    statusLabel->setText(tr("WebSocket Ğ¾Ñ‚ĞºĞ»ÑÑ‡ĞµĞ½"));
    isConnected = false;
    updateConnectionState();
    heartbeatTimer->stop();
    
    addMessageToChat(tr("Ğ¡Ğ¸ÑÑ‚ĞµĞ¼Ğ°"), tr("WebSocket ÑĞ¾ĞµĞ´Ğ¸Ğ½ĞµĞ½Ğ¸Ğµ Ğ¿Ğ¾Ñ‚ĞµÑ€ÑĞ½Ğ¾"), QDateTime::currentDateTime(), false);
    
    // ĞŸĞ¾Ğ¿Ñ‹Ñ‚ĞºĞ° Ğ¿ĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ
    if (appSettings->autoReconnect() && reconnectAttempts < 5) {
        reconnectAttempts++;
        statusLabel->setText(tr("ĞŸĞ¾Ğ¿Ñ‹Ñ‚ĞºĞ° Ğ¿ĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ... (%1)").arg(reconnectAttempts));
        reconnectTimer->start();
    }
}

void MainWindow::onWebSocketError(QAbstractSocket::SocketError error)
{
    QString errorMsg = webSocket->errorString();
    showError(tr("WebSocket Ğ¾ÑˆĞ¸Ğ±ĞºĞ°: %1").arg(errorMsg));
    statusLabel->setText(tr("WebSocket Ğ¾ÑˆĞ¸Ğ±ĞºĞ°: %1").arg(errorMsg));
}

void MainWindow::onWebSocketTextMessageReceived(const QString &message)
{
    QJsonDocument doc = QJsonDocument::fromJson(message.toUtf8());
    if (!doc.isNull()) {
        QJsonObject jsonObj = doc.object();
        handleIncomingMessage(jsonObj);
    }
}

void MainWindow::handleIncomingMessage(const QJsonObject &jsonObj)
{
    QString type = jsonObj["type"].toString();
    
    if (type == "new_message" || type == "history_message") {
        QString user = jsonObj["user"].toString();
        QString text = jsonObj["text"].toString();
        QString timestampStr = jsonObj["timestamp"].toString();
        
        QDateTime timestamp = QDateTime::fromString(timestampStr, Qt::ISODateWithMs);
        if (!timestamp.isValid()) {
            timestamp = QDateTime::currentDateTime();
        }
        
        addMessageToChat(user, text, timestamp, user == currentUsername);
        
        // ĞŸĞ¾ĞºĞ°Ğ· ÑƒĞ²ĞµĞ´Ğ¾Ğ¼Ğ»ĞµĞ½Ğ¸Ñ
        if (notificationsEnabled && user != currentUsername) {
            showNotification(tr("ĞĞ¾Ğ²Ğ¾Ğµ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğµ Ğ¾Ñ‚ %1").arg(user), text);
        }
    } else if (type == "auth_success") {
        statusLabel->setText(tr("ĞÑƒÑ‚ĞµĞ½Ñ‚Ğ¸Ñ„Ğ¸ĞºĞ°Ñ†Ğ¸Ñ ÑƒÑĞ¿ĞµÑˆĞ½Ğ°"));
        reconnectAttempts = 0; // Ğ¡Ğ±Ñ€Ğ¾Ñ Ğ¿Ğ¾Ğ¿Ñ‹Ñ‚Ğ¾Ğº Ğ¿ĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ Ğ¿Ğ¾ÑĞ»Ğµ ÑƒÑĞ¿ĞµÑˆĞ½Ğ¾Ğ¹ Ğ°ÑƒÑ‚ĞµĞ½Ñ‚Ğ¸Ñ„Ğ¸ĞºĞ°Ñ†Ğ¸Ğ¸
    } else if (type == "auth_error") {
        QString errorMsg = jsonObj["message"].toString();
        showError(tr("ĞÑˆĞ¸Ğ±ĞºĞ° Ğ°ÑƒÑ‚ĞµĞ½Ñ‚Ğ¸Ñ„Ğ¸ĞºĞ°Ñ†Ğ¸Ğ¸: %1").arg(errorMsg));
    } else if (type == "error") {
        QString errorMsg = jsonObj["message"].toString();
        showError(tr("ĞÑˆĞ¸Ğ±ĞºĞ° Ğ¾Ñ‚ ÑĞµÑ€Ğ²ĞµÑ€Ğ°: %1").arg(errorMsg));
    } else if (type == "rate_limit_exceeded") {
        QString errorMsg = jsonObj["message"].toString();
        showError(tr("ĞŸÑ€ĞµĞ²Ñ‹ÑˆĞµĞ½ Ğ»Ğ¸Ğ¼Ğ¸Ñ‚ Ğ·Ğ°Ğ¿Ñ€Ğ¾ÑĞ¾Ğ²: %1").arg(errorMsg));
    }
}

void MainWindow::addMessageToChat(const QString &user, const QString &text, const QDateTime &timestamp, bool isOwn)
{
    MessageItem item;
    item.user = user;
    item.text = text;
    item.timestamp = timestamp;
    item.isOwn = isOwn;
    
    chatModel->addMessage(item);
    
    // ĞŸÑ€Ğ¾ĞºÑ€ÑƒÑ‚ĞºĞ° Ğº Ğ¿Ğ¾ÑĞ»ĞµĞ´Ğ½ĞµĞ¼Ñƒ ÑĞ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ñ
    QModelIndex lastIndex = chatModel->index(chatModel->rowCount() - 1, 0);
    ui->chatView->scrollTo(lastIndex, QAbstractItemView::PositionAtBottom);
}

void MainWindow::sendAuthRequest()
{
    QJsonObject authMsg;
    authMsg["type"] = "auth";
    authMsg["username"] = currentUsername;
    authMsg["password"] = "temp_password"; // Ğ’ Ñ€ĞµĞ°Ğ»ÑŒĞ½Ğ¾Ğ¼ Ğ¿Ñ€Ğ¸Ğ»Ğ¾Ğ¶ĞµĞ½Ğ¸Ğ¸ Ğ½ÑƒĞ¶Ğ½Ğ¾ Ğ·Ğ°Ğ¿Ñ€Ğ°ÑˆĞ¸Ğ²Ğ°Ñ‚ÑŒ Ğ¿Ğ°Ñ€Ğ¾Ğ»ÑŒ
    
    QString authString = QJsonDocument(authMsg).toJson(QJsonDocument::Compact);
    
    if (useWebSocket) {
        webSocket->sendTextMessage(authString);
    } else {
        tcpSocket->write(authString.toUtf8() + "\n");
    }
}

void MainWindow::sendHeartbeat()
{
    if (isConnected) {
        QJsonObject heartbeatMsg;
        heartbeatMsg["type"] = "heartbeat";
        heartbeatMsg["timestamp"] = QDateTime::currentDateTimeUtc().toString(Qt::ISODateWithMs);
        
        QString heartbeatString = QJsonDocument(heartbeatMsg).toJson(QJsonDocument::Compact);
        
        if (useWebSocket) {
            webSocket->sendTextMessage(heartbeatString);
        } else {
            tcpSocket->write(heartbeatString.toUtf8() + "\n");
        }
    }
}

void MainWindow::updateConnectionState()
{
    bool connected = isConnected;
    
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ Ñ‚ĞµĞºÑÑ‚Ğ° ĞºĞ½Ğ¾Ğ¿ĞºĞ¸ Ğ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ñ
    if (connected) {
        ui->connectButton->setText(tr("ĞÑ‚ĞºĞ»ÑÑ‡Ğ¸Ñ‚ÑŒÑÑ"));
        ui->connectButton->setStyleSheet("QPushButton { background-color: #f44336; color: white; }");
        connectionStatusLabel->setText(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½"));
    } else {
        ui->connectButton->setText(tr("ĞŸĞ¾Ğ´ĞºĞ»ÑÑ‡Ğ¸Ñ‚ÑŒÑÑ"));
        ui->connectButton->setStyleSheet("");
        connectionStatusLabel->setText(tr("ĞÑ‚ĞºĞ»ÑÑ‡ĞµĞ½"));
    }
    
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ Ğ´Ğ¾ÑÑ‚ÑƒĞ¿Ğ½Ğ¾ÑÑ‚Ğ¸ ÑĞ»ĞµĞ¼ĞµĞ½Ñ‚Ğ¾Ğ² ÑƒĞ¿Ñ€Ğ°Ğ²Ğ»ĞµĞ½Ğ¸Ñ
    ui->messageInput->setEnabled(connected);
    ui->sendButton->setEnabled(connected);
    ui->usernameEdit->setEnabled(!connected);
    ui->serverEdit->setEnabled(!connected);
    ui->portSpinBox->setEnabled(!connected);
    ui->protocolCombo->setEnabled(!connected);
    ui->sslCheckBox->setEnabled(!connected);
    
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ Ğ´ĞµĞ¹ÑÑ‚Ğ²Ğ¸Ğ¹ Ğ² Ñ‚Ñ€ĞµĞµ
    if (trayIcon) {
        connectAction->setEnabled(!connected);
        disconnectAction->setEnabled(connected);
    }
}

void MainWindow::showError(const QString &error)
{
    QMessageBox::critical(this, tr("ĞÑˆĞ¸Ğ±ĞºĞ°"), error);
    statusLabel->setText(tr("ĞÑˆĞ¸Ğ±ĞºĞ°: %1").arg(error));
}

void MainWindow::showNotification(const QString &title, const QString &message)
{
    if (notificationsEnabled && trayIcon) {
        trayIcon->showMessage(title, message, QSystemTrayIcon::Information, 5000);
    }
}

void MainWindow::showTrayMenu()
{
    if (trayMenu) {
        trayMenu->popup(QCursor::pos());
    }
}

void MainWindow::trayActivated(QSystemTrayIcon::ActivationReason reason)
{
    switch (reason) {
    case QSystemTrayIcon::Trigger:
        if (isMinimized()) {
            showNormal();
            raise();
            activateWindow();
        } else {
            showMinimized();
        }
        break;
    case QSystemTrayIcon::DoubleClick:
        showNormal();
        raise();
        activateWindow();
        break;
    default:
        break;
    }
}

void MainWindow::showSettingsDialog()
{
    // Ğ—Ğ´ĞµÑÑŒ Ğ±ÑƒĞ´ĞµÑ‚ Ğ²Ñ‹Ğ·Ğ¾Ğ² Ğ´Ğ¸Ğ°Ğ»Ğ¾Ğ³Ğ° Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº
    // Ğ”Ğ»Ñ ÑƒĞ¿Ñ€Ğ¾Ñ‰ĞµĞ½Ğ¸Ñ Ğ² ÑÑ‚Ğ¾Ğ¼ Ğ¿Ñ€Ğ¸Ğ¼ĞµÑ€Ğµ Ğ¿Ñ€Ğ¾ÑÑ‚Ğ¾ Ğ¿Ğ¾ĞºĞ°Ğ·Ñ‹Ğ²Ğ°ĞµĞ¼ Ğ²ĞºĞ»Ğ°Ğ´ĞºÑƒ Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº
    if (ui->tabWidget->indexOf(ui->settingsTab) != -1) {
        ui->tabWidget->setCurrentWidget(ui->settingsTab);
    }
}

void MainWindow::applySettings()
{
    // ĞŸÑ€Ğ¸Ğ¼ĞµĞ½ĞµĞ½Ğ¸Ğµ Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº Ğ¸Ğ· Ğ¸Ğ½Ñ‚ĞµÑ€Ñ„ĞµĞ¹ÑĞ°
    currentUsername = ui->usernameEdit->text();
    currentServer = ui->serverEdit->text();
    currentPort = ui->portSpinBox->value();
    useWebSocket = ui->protocolCombo->currentIndex() == 1; // 0 - TCP, 1 - WebSocket
    
    // Ğ¡Ğ¾Ñ…Ñ€Ğ°Ğ½ĞµĞ½Ğ¸Ğµ Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº
    saveSettings();
    
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ Ğ·Ğ°Ğ³Ğ¾Ğ»Ğ¾Ğ²ĞºĞ° Ğ¾ĞºĞ½Ğ°
    setWindowTitle(tr("ĞœĞµÑÑĞµĞ½Ğ´Ğ¶ĞµÑ€ - ") + currentServer + ":" + QString::number(currentPort));
}

void MainWindow::onUsernameChanged(const QString &username)
{
    currentUsername = username;
}

void MainWindow::onServerAddressChanged(const QString &address)
{
    currentServer = address;
}

void MainWindow::onPortChanged(int port)
{
    currentPort = port;
}

void MainWindow::onUseSslChanged(bool useSsl)
{
    Q_UNUSED(useSsl);
    // ĞŸĞµÑ€ĞµĞ¿Ğ¾Ğ´ĞºĞ»ÑÑ‡ĞµĞ½Ğ¸Ğµ Ğ¿Ñ€Ğ¸ Ğ¸Ğ·Ğ¼ĞµĞ½ĞµĞ½Ğ¸Ğ¸ SSL Ğ½Ğ°ÑÑ‚Ñ€Ğ¾ĞµĞº
    if (isConnected) {
        disconnectFromServer();
        QTimer::singleShot(1000, this, &MainWindow::connectToServer);
    }
}

void MainWindow::closeEvent(QCloseEvent *event)
{
    if (trayIcon && trayIcon->isVisible()) {
        hide();
        event->ignore();
    } else {
        if (isConnected) {
            disconnectFromServer();
        }
        event->accept();
    }
}

void MainWindow::updateStatusBar()
{
    // ĞĞ±Ğ½Ğ¾Ğ²Ğ»ĞµĞ½Ğ¸Ğµ Ğ¸Ğ½Ñ„Ğ¾Ñ€Ğ¼Ğ°Ñ†Ğ¸Ğ¸ Ğ² ÑÑ‚Ñ€Ğ¾ĞºĞµ ÑĞ¾ÑÑ‚Ğ¾ÑĞ½Ğ¸Ñ
    QString info = tr("ĞŸĞ¾Ğ»ÑŒĞ·Ğ¾Ğ²Ğ°Ñ‚ĞµĞ»ĞµĞ¹ Ğ¾Ğ½Ğ»Ğ°Ğ¹Ğ½: %1 | Ğ¡Ğ¾Ğ¾Ğ±Ñ‰ĞµĞ½Ğ¸Ğ¹: %2")
                       .arg(chatModel->onlineUsersCount())
                       .arg(chatModel->messageCount());
    statusLabel->setText(info);
}